From 424bb16b7df737b50a4bb0ef7e01a795317df1d4 Mon Sep 17 00:00:00 2001
From: Paul Harrison <pharrison@mythtv.org>
Date: Sun, 23 Feb 2014 13:14:58 +0000
Subject: [PATCH 4/8] MythMusic: improve the handling of unavailable tracks in
 the player

* add a MusicPlayerEvent::TrackUnavailableEvent which is sent when trying to
  playback a track that cannot be found.

* skip unavailable tracks and show a notification to let the user know we can't
  play the track.
---
 mythplugins/mythmusic/mythmusic/musicplayer.cpp |   36 ++++++++++++++++++++++-
 mythplugins/mythmusic/mythmusic/musicplayer.h   |    2 ++
 2 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/mythplugins/mythmusic/mythmusic/musicplayer.cpp b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
index b7ba3d6..175034b 100644
--- a/mythplugins/mythmusic/mythmusic/musicplayer.cpp
+++ b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
@@ -42,6 +42,7 @@ QEvent::Type MusicPlayerEvent::TrackChangeEvent = (QEvent::Type) QEvent::registe
 QEvent::Type MusicPlayerEvent::VolumeChangeEvent = (QEvent::Type) QEvent::registerEventType();
 QEvent::Type MusicPlayerEvent::TrackAddedEvent = (QEvent::Type) QEvent::registerEventType();
 QEvent::Type MusicPlayerEvent::TrackRemovedEvent = (QEvent::Type) QEvent::registerEventType();
+QEvent::Type MusicPlayerEvent::TrackUnavailableEvent = (QEvent::Type) QEvent::registerEventType();
 QEvent::Type MusicPlayerEvent::AllTracksRemovedEvent = (QEvent::Type) QEvent::registerEventType();
 QEvent::Type MusicPlayerEvent::MetadataChangedEvent = (QEvent::Type) QEvent::registerEventType();
 QEvent::Type MusicPlayerEvent::TrackStatsChangedEvent = (QEvent::Type) QEvent::registerEventType();
@@ -321,11 +322,35 @@ void MusicPlayer::pause(void)
 
 void MusicPlayer::play(void)
 {
+    stopDecoder();
+
     MusicMetadata *meta = getCurrentMetadata();
     if (!meta)
         return;
 
-    stopDecoder();
+    if (meta->Filename() == METADATA_INVALID_FILENAME)
+    {
+        // put an upper limit on the number of consecutive track unavailable errors
+        if (m_errorCount >= 1000)
+        {
+            ShowOkPopup(tr("Got to many track unavailable errors. Maybe the host with the music on is off-line?"));
+            stop(true);
+            m_errorCount = 0;
+            return;
+        }
+
+        if (m_errorCount < 5)
+        {
+            MythErrorNotification n(tr("Track Unavailable"), tr("MythMusic"), QString("Cannot find file '%1'").arg(meta->Filename(false)));
+            GetNotificationCenter()->Queue(n);
+        }
+
+        m_errorCount++;
+
+        sendTrackUnavailableEvent(meta->ID());
+        next();
+        return;
+    }
 
     // Notify others that we are about to play
     gCoreContext->WantingPlayback(this);
@@ -1326,6 +1351,12 @@ void MusicPlayer::sendAlbumArtChangedEvent(int trackID)
     dispatch(me);
 }
 
+void MusicPlayer::sendTrackUnavailableEvent(int trackID)
+{
+    MusicPlayerEvent me(MusicPlayerEvent::TrackUnavailableEvent, trackID);
+    dispatch(me);
+}
+
 void MusicPlayer::sendCDChangedEvent(void)
 {
     MusicPlayerEvent me(MusicPlayerEvent::CDChangedEvent, -1);
@@ -1478,6 +1509,9 @@ void MusicPlayer::setupDecoderHandler(void)
 
 void MusicPlayer::decoderHandlerReady(void)
 {
+    if (!getDecoder())
+        return;
+
     LOG(VB_PLAYBACK, LOG_INFO, QString ("decoder handler is ready, decoding %1")
             .arg(getDecoder()->getFilename()));
 
diff --git a/mythplugins/mythmusic/mythmusic/musicplayer.h b/mythplugins/mythmusic/mythmusic/musicplayer.h
index b0ec7b0..e64d898 100644
--- a/mythplugins/mythmusic/mythmusic/musicplayer.h
+++ b/mythplugins/mythmusic/mythmusic/musicplayer.h
@@ -38,6 +38,7 @@ class MusicPlayerEvent : public MythEvent
         static Type VolumeChangeEvent;
         static Type TrackAddedEvent;
         static Type TrackRemovedEvent;
+        static Type TrackUnavailableEvent;
         static Type AllTracksRemovedEvent;
         static Type MetadataChangedEvent;
         static Type TrackStatsChangedEvent;
@@ -141,6 +142,7 @@ class MusicPlayer : public QObject, public MythObservable
     void         sendMetadataChangedEvent(int trackID);
     void         sendTrackStatsChangedEvent(int trackID);
     void         sendAlbumArtChangedEvent(int trackID);
+    void         sendTrackUnavailableEvent(int trackID);
     void         sendCDChangedEvent(void);
 
     void         toMap(InfoMap &infoMap);
-- 
1.7.10

